File: data-structures/non-linear/binary-tree/heap/max-heap/MaxHeap.py
class MaxHeap:
    def __init__(self, arr=None):
        self.heap = []
        if arr:
            self.build_heap(arr)

    def _left(self, i):
        return 2 * i + 1

    def _right(self, i):
        return 2 * i + 2

    def _parent(self, i):
        return (i - 1) // 2

    def _heapify(self, i):
        n = len(self.heap)
        largest = i
        l, r = self._left(i), self._right(i)

        if l < n and self.heap[l] > self.heap[largest]:
            largest = l
        if r < n and self.heap[r] > self.heap[largest]:
            largest = r

        if largest != i:
            self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
            self._heapify(largest)

    def build_heap(self, arr):
        self.heap = arr[:]
        n = len(self.heap)
        for i in range(n // 2 - 1, -1, -1):
            self._heapify(i)

    def extract_max(self):
        if not self.heap:
            return None
        if len(self.heap) == 1:
            return self.heap.pop()

        root = self.heap[0]
        self.heap[0] = self.heap.pop()
        self._heapify(0)
        return root

    def insert(self, key):
        self.heap.append(key)
        i = len(self.heap) - 1
        while i > 0 and self.heap[self._parent(i)] < self.heap[i]:
            self.heap[self._parent(i)], self.heap[i] = (
                self.heap[i],
                self.heap[self._parent(i)],
            )
            i = self._parent(i)

    def get_max(self):
        return None if not self.heap else self.heap[0]

    def size(self):
        return len(self.heap)

    def display(self):
        print(self.heap)

File: data-structures/non-linear/binary-tree/heap/max-heap/MaxHeap.ts
class MaxHeap {
  private heap: Array<number>;

  private getParent(index: number): number {
    return (index - 1) / 2;
  }

  private getLeftChild(index: number): number {
    return index * 2 + 1;
  }

  private getRightChild(index: number): number {
    return index * 2 + 2;
  }

  private upHeap(index: number): void {
    let temp: number;
    while (index > 0 && this.heap[index] > this.heap[this.getParent(index)]) {
      temp = this.heap[index];
      this.heap[index] = this.heap[this.getParent(index)];
      this.heap[this.getParent(index)] = temp;
      index = this.getParent(index);
    }
  }

  private downHeap(index: number): void {
    let smallest: number = index;
    const leftChild: number = this.getLeftChild(index);
    const rightChild: number = this.getRightChild(index);

    if (
      leftChild < this.heap.length &&
      this.heap[smallest] < this.heap[leftChild]
    ) {
      smallest = leftChild;
    }

    if (
      rightChild < this.heap.length &&
      this.heap[smallest] < this.heap[rightChild]
    ) {
      smallest = leftChild;
    }

    if (index != smallest) {
      const temp: number = this.heap[index];
      this.heap[index] = this.heap[smallest];
      this.heap[smallest] = temp;
      this.upHeap(smallest);
    }
  }

  public isEmpty(): boolean {
    return this.heap.length == 0;
  }

  public size(): number {
    return this.heap.length;
  }

  public insert(key: number): void {
    this.heap.push(key);
    this.upHeap(this.heap.length - 1);
  }

  public getMax(): number {
    if (this.isEmpty()) return -1;

    return this.heap[0];
  }

  public extractMax(): number {
    if (this.isEmpty()) return -1;

    const min: number = this.heap[0];

    //@ts-expect-error If the heap is empty the function terminates early, array.pop returns undefined when empty array
    this.heap[0] = this.heap.pop();

    if (!this.isEmpty()) {
      this.downHeap(0);
    }

    return min;
  }

  public buildHeap(heapArray: Array<number>): void {
    this.heap = heapArray;

    for (let i = this.getParent(this.heap.length - 1); i >= 0; i--) {
      this.downHeap(i);
    }
  }

  public printHeap(): void {
    for (const x of this.heap) {
      console.log(`${x}`);
    }
  }
}

export { MaxHeap };
